File: acme-java-springboot/src/main/java/acme/service/TransactionService.java
package acme.service;

import acme.dto.TransactionRequest;
import acme.dto.TransactionResponse;
import acme.entity.Account;
import acme.entity.Transaction;
import acme.repository.AccountRepository;
import acme.repository.TransactionRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
public class TransactionService {

    private static final Logger log = LoggerFactory.getLogger(TransactionService.class);

    private final AccountRepository accountRepository;
    private final TransactionRepository transactionRepository;

    public TransactionService(AccountRepository accountRepository, TransactionRepository transactionRepository) {
        this.accountRepository = accountRepository;
        this.transactionRepository = transactionRepository;
    }

    @Transactional
    public TransactionResponse performTransaction(TransactionRequest request) {
        // validate accounts exist
        Account fromAccount = accountRepository.findById(request.from())
            .orElseThrow(() -> new IllegalArgumentException("From account not found: " + request.from()));

        Account toAccount = accountRepository.findById(request.to())
            .orElseThrow(() -> new IllegalArgumentException("To account not found: " + request.to()));

        // validate amount
        if (request.amount() <= 0) {
            throw new IllegalArgumentException("Transaction amount must be positive");
        }

        // validate sufficient funds
        if (fromAccount.amount() < request.amount()) {
            throw new IllegalArgumentException("Insufficient funds in account: " + request.from());
        }

        if (fromAccount.withdrawalLimit() != -1L) {
            // Check transaction limit for last 10 transactions
            double totalLast10Transactions = transactionRepository.sumLastNTransactionsByAccountId(
                request.from(),
                10
            );

            if (totalLast10Transactions > fromAccount.withdrawalLimit()) {
                throw new IllegalArgumentException(
                    "Transaction would exceed limit for last 10 transactions. " +
                    "Current total of last 10 transactions: " + totalLast10Transactions + ", " +
                    "Limit: " + fromAccount.withdrawalLimit() + ", " +
                    "Attempted amount: " + request.amount()
                );
            }
        }

        // perform the transaction
        Account updatedFromAccount = new Account(
            fromAccount.accountId(),
            fromAccount.bankId(),
            fromAccount.routingNumber(),
            fromAccount.amount() - request.amount(),
            fromAccount.withdrawalLimit(),
            LocalDateTime.now(),
            fromAccount.version()
        );

        Account updatedToAccount = new Account(
            toAccount.accountId(),
            toAccount.bankId(),
            toAccount.routingNumber(),
            toAccount.amount() + request.amount(),
            toAccount.withdrawalLimit(),
            LocalDateTime.now(),
            toAccount.version()
        );

        // save updated accounts
        accountRepository.save(updatedFromAccount);
        accountRepository.save(updatedToAccount);

        // save transaction record
        Transaction transaction = new Transaction(
            null,   // let the database generate the ID
            request.from(),
            request.to(),
            request.amount(),
            LocalDateTime.now(),
            "SUCCESS"
        );
        Transaction savedTransaction = transactionRepository.save(transaction);

        log.info("Transaction performed: {}", savedTransaction);

        UUID transactionId = savedTransaction.transactionId();
        if (transactionId == null) {
            throw new IllegalStateException("Failed to save transaction - no ID generated");
        }

        return new TransactionResponse(
            transactionId,
            request.from(),
            request.to(),
            request.amount()
        );
    }

    public List<Transaction> getTransactionHistory(UUID accountId) {
        return transactionRepository.findByAccountId(accountId);
    }
}

File: acme-springboot/src/main/kotlin/acme/service/TransactionService.kt
package acme.service

import acme.dto.TransactionRequest
import acme.dto.TransactionResponse
import acme.entity.Transaction
import acme.repository.AccountRepository
import acme.repository.TransactionRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import java.util.UUID

@Service
class TransactionService(
    private val accountRepository: AccountRepository,
    private val transactionRepository: TransactionRepository
) {

    private val log = LoggerFactory.getLogger(this::class.java)
    
    @Transactional
    fun performTransaction(request: TransactionRequest): TransactionResponse {
        // validate accounts exist
        val fromAccount = accountRepository.findById(request.from)
            .orElseThrow { IllegalArgumentException("From account not found: ${request.from}") }
        
        val toAccount = accountRepository.findById(request.to)
            .orElseThrow { IllegalArgumentException("To account not found: ${request.to}") }
        
        // validate amount
        if (request.amount <= 0) {
            throw IllegalArgumentException("Transaction amount must be positive")
        }
        
        // validate sufficient funds
        if (fromAccount.amount < request.amount) {
            throw IllegalArgumentException("Insufficient funds in account: ${request.from}")
        }

        if (fromAccount.withdrawalLimit != -1L) {
            // Check transaction limit for last 10 transactions
            val totalLast10Transactions = transactionRepository.sumLastNTransactionsByAccountId(
                request.from,
                10
            )

            if (totalLast10Transactions > fromAccount.withdrawalLimit) {
                throw IllegalArgumentException(
                    "Transaction would exceed limit for last 10 transactions. " +
                            "Current total of last 10 transactions: $totalLast10Transactions, " +
                            "Limit: ${fromAccount.withdrawalLimit}, " +
                            "Attempted amount: ${request.amount}"
                )
            }
        }
        
        // perform the transaction
        val updatedFromAccount = fromAccount.copy(
            amount = fromAccount.amount - request.amount,
            lastChange = LocalDateTime.now()
        )
        
        val updatedToAccount = toAccount.copy(
            amount = toAccount.amount + request.amount,
            lastChange = LocalDateTime.now()
        )
        
        // save updated accounts
        accountRepository.save(updatedFromAccount)
        accountRepository.save(updatedToAccount)
        
        // save transaction record
        val transaction = Transaction(
            transactionId = null,   // let the database generate the ID
            fromAccountId = request.from,
            toAccountId = request.to,
            amount = request.amount,
            timestamp = LocalDateTime.now(),
            status = "SUCCESS"
        )
        val savedTransaction = transactionRepository.save(transaction)

        log.info("Transaction performed: $savedTransaction")

        val transactionId = savedTransaction.transactionId 
            ?: throw IllegalStateException("Failed to save transaction - no ID generated")

        return TransactionResponse(
            transactionId = transactionId,
            fromAccountId = request.from,
            toAccountId = request.to,
            amount = request.amount
        )
    }

    fun getTransactionHistory(accountId: UUID): List<Transaction> {
        return transactionRepository.findByAccountId(accountId)
    }
}
